packages/cli/src/build.ts

```diff
@@ -51,157 +58,153 @@ export function build({
-    outputCSSNameTemplate = '[filename].css',
-    outputSources = false,
-    useSourceNamespace = false,
-    injectCSSRequest = false,
-    optimize = false,
-    minify = false
+    log: (...args: string[]) => void,
+    generatorPath?: string
 ) {
-    const outSrcPath = join(fullOutDir, filePath.replace(fullSrcDir, ''));
-    const outPath = outSrcPath + '.js';
-    const fileDirectory = dirname(filePath);
-    const outDirPath = dirname(outPath);
-    const cssAssetFilename = nameTemplate(outputCSSNameTemplate, {
-        filename: basename(outSrcPath, '.st.css'),
-    });
-    const cssAssetOutPath = join(dirname(outSrcPath), cssAssetFilename);
-
-    log('[Build]', filePath + ' --> ' + outPath);
-    tryRun(() => ensureDirectory(outDirPath, fs), `Ensure directory: ${outDirPath}`);
-    let content: string = tryRun(
-        () => fs.readFileSync(filePath).toString(),
-        `Read File Error: ${filePath}`
-    );
-    const res = stylable.transform(content, filePath);
-    const optimizer = new StylableOptimizer();
-    if (optimize) {
-        optimizer.optimize(
-            {
-                removeComments: true,
-                removeEmptyNodes: true,
-                removeStylableDirectives: true,
-                classNameOptimizations: false,
-                removeUnusedComponents: false,
-            },
-            res,
-            {}
-        );
-    }
-    handleDiagnostics(res, diagnosticsMsg, filePath);
-    // st.css
-    if (outputSources) {
-        if (useSourceNamespace && !content.includes('st-namespace-reference')) {
-            const relativePathToSource = relative(dirname(outSrcPath), filePath).replace(
-                /\\/gm,
-                '/'
-            );
-            const srcNamespaceAnnotation = `/* st-namespace-reference="${relativePathToSource}" */\n`;
-            content = srcNamespaceAnnotation + content;
-        }
+    const generatorModule: { Generator: typeof Generator } = generatorPath
+        ? require(generatorPath)
+        : require('./base-generator');
+    return new generatorModule.Generator(stylable, log);
+}
 
-        log('[Build]', 'output .st.css source');
-        tryRun(() => fs.writeFileSync(outSrcPath, content), `Write File Error: ${outSrcPath}`);
-    }
-    // st.css.js
-    moduleFormats.forEach((format) => {
-        log('[Build]', 'moduleFormat', format);
-        const code = tryRun(
-            () =>
-                createModuleSource(
-                    res,
-                    format,
-                    includeCSSInJS,
-                    undefined,
-                    undefined,
-                    undefined,
-                    injectCSSRequest ? [`./${cssAssetFilename}`] : [],
-                    '@stylable/runtime'
-                ),
-            `Transform Error: ${filePath}`
+function validateConfiguration(outputSources: boolean | undefined, outDir: string, srcDir: string) {
+    if (outputSources && srcDir === outDir) {
```

Comment:
this only protect overriding source files. I can probably remove this check since I now check it in build single file also. but maybe it's good to exit early